### ISSUE TITLE
TAB-character (ascii 9) is not displayed correctly for Stardict dictionaries with sametypesequence m and x

### ISSUE THREAD

**Issue by Markismus** [koreader/koreader#5837]
* KOReader version: v2019.12/v2020.01
* Device: Linux

* KOReader version: v2020.01
* Device: Kobo Aura H2O 1st gen.

#### Issue
TAB-character (ascii 9) is not displayed correctly for Stardict dictionaries with sametypesequence m and x, but is for h. (I put in 4 tabs to make it obvious.)

(In the Stardict ifo-file are a number of values for the dictionary. One line is dedicated to sametypesequence, e.g. "sametypesequence=x". m is text, x is xdxf markup text, h is html text)

sametypesequence=x and sametypesequence=m 
![Screenshot from 2020-02-09 18-44-30](https://user-images.githubusercontent.com/5269101/74109479-09c6a100-4b84-11ea-89bc-e987252cd949.png)

sametypesequence=h
![Screenshot from 2020-02-09 18-44-47](https://user-images.githubusercontent.com/5269101/74109482-10551880-4b84-11ea-8a0b-68f32636a617.png)

(Obviously, the html version stumbles over the blockquote-tags that are abundant in Oxford 2nd Edition. I was testing the 'x' and 'm' handling to see whether it gave a better result.)

#### Steps to reproduce
[Here is the sample dictionary with 7 tabs at the entry 'a'](https://markismus.stackstorage.com/s/dRgfLeoYL8xDScF)

**Comment by Markismus** [comment]
Anybody any idea what part of the code is relevant? I haven't looked at the lua-code for a few years and it kind of grew rather large...

I can easily work around it by substituting all tabs and replacing them by spaces or whatever, but I don't see why tabs should not be represented. What am I missing here?

**Comment by Frenzie** [comment]
I would start at no code at all, just sdcv. See here for the command:

https://github.com/koreader/koreader/blob/0599c440cc934c1dceb5a2fed4aebb02bf01823e/frontend/apps/reader/modules/readerdictionary.lua#L692

**Comment by Markismus** [comment]
I know what it puts out, it put it in there myself in a Stardict xml-file. That's where I found that the symbol was a tab. That's were I put multiple tabs in a row.

Here is the result in the terminal:
```
koreader-appimage]$ ./sdcv --utf8-input --utf8-output --non-interactive --data-dir data/dict a
Found 1 items, similar to a.
-->Oxford English Dictionary 2nd Ed. P1
-->a


▪ I.								†  a, a.1 (def. numeral) Obs. or dial.
[OE. án, one, of which the n began to disappear before a cons. about 1150. In the definite numeral sense, án and á, following the ordinary course of OE. long á, became in the south bef. 1300, on (oon, one), o (oo); and eventually o became obs., leaving one as the form in all positions; while an and a, pronounced lightly and indistinctly, became the ‘indefinite article.’ See next word. But in the north an (or ane) and a were written in both senses, the stress or emphasis alone distinguishing the numeral from the article.]
```

I've written a [converter for PocketBook](https://github.com/Markismus/PocketBookDic), but since I returned the device, I now adopted the script to also generate dictionaries for Stardict. I can fix whatever needs fixing in the script, but this seems to be handled in the koreader code.

**Comment by poire-z** [comment]
I guess there's just no glyph for TAB in any font :)

It can probably be cleaned up there, where we already do some cleanup:
https://github.com/koreader/koreader/blob/833fed79ffce291815ccaaa106de17eb5f1f3026/frontend/apps/reader/modules/readerdictionary.lua#L540-L578

Dunno if it was cleaned up somehow before the switch to xtext, but I don't see anything in textboxwidget.lua, rendertext.lua or ffi/freetype.lua dealing with TAB.

**Comment by poire-z** [comment]
`sametypesequence=h` is displayed by MuPDF HTML renderer, so hopefully ok :)
Others are displayed by KOReader own text rendering stuff, and it looks like there's nothing special to deal with TAB: we ask the font for the glyph, and if there's none, it's the tofu glyph.

**Comment by Markismus** [comment]
So the display of the tab character is always dealt with by the viewer/browser? As it has variable width, that would seem more logical than defining a fixed width font glyph...
And in html-mode it is handled by another interpreter.

**Comment by poire-z** [comment]
Well, we don't handle it specifically.
Either you substitute here with 1 or N spaces.
Either we add support for that in textboxwidget to substitute it with a computed number of spaces to ensure some kind of grid fitting - but it feels it would not be easy, as it's Harfbuzz that does most of the text layout work (not talking about TABS is BiDi text :)
And how does it do in justified text?

A quick search does not show there's been any discussion about the handling of tabs in Harfbuzz... I'll have to look if there's some specific handling of them in the various other projects I looked at previously.

**Comment by NiLuJe** [comment]
`s/\t/    /g`? ^^

**Comment by poire-z** [comment]
Given the use case and the screenshot, I guess that yes, a space will do :)
Just add at the end of the above `local function tidyMarkup(results) `
```diff
              -- strip all leading empty lines/spaces 
              def = def:gsub("^%s+", "") 
+             -- substitute tab with space
+             def = def:gsub("\t", " ") 
              result.definition = def 
```

Some more correct tab expansion: https://github.com/jgm/lunamark/blob/90e35a42df585838e29a9ba8b93f7aae6eb8dbf3/lunamark/util.lua#L89-L100
but I guess this is fine only with a fixed size font.
For more correct handling, i guess it should be done by TextBoxWidget that should manage tabstops, in a way like it manages text justification, by spacing/pushing words when there is a space or a tab, possibly needing to have xtext/Harfbuzz re-measure the rest of the text once a tabstop has pushed that text away and there's less width available on the line than before the tab push.
Just some thoughts, not planning to implement that, unless some other use case shows up.
Again, given the use case/screenshot, it would do more harm than a simple substitution with a space :)

**Comment by Markismus** [comment]
[From What HarfBuzz doesn't do:](https://harfbuzz.github.io/what-harfbuzz-doesnt-do.html)

> HarfBuzz can tell you how wide a shaped piece of text is, which is useful input to a justification algorithm, but it knows nothing about paragraphs, lines or line lengths. Nor will it adjust the space between words to fit them proportionally into a line. 

So I understand that HarfBuzz doesn't work for adjustable spacing and I understand that that would have to be implemented, but foremost it seems odd that a character such as a tab is given to a font for expression if there isn't a possible glyph for it. 

How are the others from the group handled?

>8 | 08 | 010 | 0001000 | BS | backspace
>9 | 09 | 011 | 0001001 | HT | horizontal tab
>10 | 0A | 012 | 0001010 | LF | line feed
>11 | 0B | 013 | 0001011 | VT | vertical tab
>12 | 0C | 014 | 0001100 | FF | form feed
>13 | 0D | 015 | 0001101 | CR | carriage return

**Comment by poire-z** [comment]
> From What HarfBuzz doesn't do

And what HB doesn't do, we had to do it ourselves :) And we did with https://github.com/koreader/koreader-base/pull/1010 and https://github.com/koreader/koreader/commit/e699a1ee

I guess the ones you mentionned should obviously be handled by our code that is using harfbuzz, like `\n` and `\t`, so we give only single lines to HB. It can't know what we want to do with TABs.
The other you mentionned are sufficiently rare and I wouldn't know what to do with them :) So, if you have them in a bit of text, it's as good they are shown with that tofu glyph, so you know you have some crap that might cause some issues (e.g. in TextEditor).

For info, Unicode and HB have their own list of "ignorable" codepoints:
https://github.com/harfbuzz/harfbuzz/blob/4dc87365d79d9cb0ad85b351b6afe978c3abee3c/src/hb-unicode.hh#L131-L167

**Comment by poire-z** [comment]
> it seems odd that a character such as a tab is given to a font for expression if there isn't a possible glyph for it

Some fonts may have a gyph for it, which could look like `[TAB]`. Some fonts have that kind of glyph for when they have not the real glyph:
![image](https://user-images.githubusercontent.com/24273478/74134511-fdc1fa00-4bea-11ea-9bfc-dbe68f81d6ad.png)

**Comment by NiLuJe** [comment]
Pango *may* have code to handle some of those escape codes, but it's
entirely out of harfbuzz's scope to do anything with those.

**Comment by Markismus** [comment]
There is a sametypesetting=g for Pango markup dictionaries in Stardict. However, I don't think you meant that Pango is already implemented in koreader, right?

**Comment by NiLuJe** [comment]
Yep, we have a @poire-z instead ^^.

**Comment by poire-z** [comment]
Ok, let's have this @poire-z fixed by the above 2 PRs :)
(I initially thought: _why bother, we have so little chances to meet TABs with our kind of input_ - and endup with _why bother if I didn't get that exactly right and pretty, we have so little chances to meet TABs with our kind of input_ :)

Yes, Pango is a big text-layout library, with many dependencies. We did our own library, that just fits with how our frontend do stuff. (Some info about libraries we ended up using or not in my text layout notes at https://github.com/koreader/crengine/issues/307).
And I guess Pango has its own markup language, that stardict is aware of.
